File: Sources/OpenMCMetal/Tally/Statistics.swift
import Foundation

// MARK: - TallyStatistics (namespace alias)
typealias TallyStatistics = Statistics

// MARK: - Statistics

/// Statistical utilities for Monte Carlo tally post-processing.
///
/// Uses Welford's online algorithm for numerically stable mean and variance
/// accumulation, avoiding catastrophic cancellation for large batch counts.
enum Statistics {

    // MARK: - Core Welford Accumulation

    /// Compute (mean, population std-dev-of-mean, relative error) for each bin
    /// across all batches using Welford's online algorithm.
    ///
    /// - Parameters:
    ///   - batchValues: Array of batches; each batch is an array of `size` bin values.
    ///   - size: Number of tally bins (must equal `batchValues[i].count` for all i).
    /// - Returns: Array of `(mean, stdDev, relError)` tuples, one per bin.
    static func computeStats(
        batchValues: [[Double]]
    ) -> [(mean: Double, stdDev: Double, relError: Double)] {
        guard let first = batchValues.first else { return [] }
        let n         = first.count
        let numBatches = batchValues.count

        var results: [(mean: Double, stdDev: Double, relError: Double)] = []
        results.reserveCapacity(n)

        for i in 0..<n {
            var mean = 0.0
            var m2   = 0.0

            for b in 0..<numBatches {
                let x      = batchValues[b][i]
                let delta  = x - mean
                mean      += delta / Double(b + 1)
                let delta2 = x - mean
                m2        += delta * delta2
            }

            // Sample variance of the batch mean = s² / N
            let variance = numBatches > 1 ? m2 / Double(numBatches - 1) : 0.0
            let stdDev   = sqrt(variance / Double(numBatches))   // std dev of mean
            let relError = mean != 0.0 ? stdDev / abs(mean) : 0.0

            results.append((mean: mean, stdDev: stdDev, relError: relError))
        }

        return results
    }

    // MARK: - k-effective Statistics

    /// Compute mean, std-dev-of-mean, and 95 % confidence interval for
    /// a series of per-batch k-effective estimates using Welford's algorithm.
    ///
    /// - Parameter batchKEff: Per-batch k-eff values (active batches only).
    /// - Returns: `(mean, stdDev, ci95)` where `ci95 = (lower, upper)`.
    static func kEffStatistics(
        batchKEff: [Double]
    ) -> (mean: Double, stdDev: Double, ci95: (Double, Double)) {
        let n = batchKEff.count
        guard n > 0 else { return (0, 0, (0, 0)) }

        var mean = 0.0
        var m2   = 0.0

        for i in 0..<n {
            let delta  = batchKEff[i] - mean
            mean      += delta / Double(i + 1)
            let delta2 = batchKEff[i] - mean
            m2        += delta * delta2
        }

        let variance = n > 1 ? m2 / Double(n - 1) : 0.0
        let stdDev   = sqrt(variance / Double(n))   // std dev of mean
        let ci95     = (mean - 1.96 * stdDev, mean + 1.96 * stdDev)

        return (mean: mean, stdDev: stdDev, ci95: ci95)
    }

    // MARK: - Figure of Merit

    /// FOM = 1 / (R² * T), where R is relative error and T is wall time in seconds.
    /// Higher FOM indicates better computational efficiency.
    ///
    /// - Parameters:
    ///   - relError: Relative error of the tally of interest.
    ///   - wallTime: Total simulation wall time in seconds.
    /// - Returns: Figure of merit, or 0 if inputs are degenerate.
    static func figureOfMerit(relError: Double, wallTime: Double) -> Double {
        guard relError > 0.0, wallTime > 0.0 else { return 0.0 }
        return 1.0 / (relError * relError * wallTime)
    }

    // MARK: - Batch Mean Helper

    /// Compute the mean across batches for each bin.
    /// Lighter-weight than `computeStats` when only means are needed.
    ///
    /// - Parameters:
    ///   - batchValues: Array of batches; each batch is an array of bin values.
    ///   - size: Expected number of bins per batch.
    /// - Returns: Array of means, length `size`.
    static func batchMeans(batchValues: [[Double]], size: Int) -> [Double] {
        guard !batchValues.isEmpty, size > 0 else {
            return [Double](repeating: 0, count: size)
        }
        var sums = [Double](repeating: 0, count: size)
        for batch in batchValues {
            for i in 0..<min(size, batch.count) {
                sums[i] += batch[i]
            }
        }
        let n = Double(batchValues.count)
        return sums.map { $0 / n }
    }
}

File: python/src/openmc_metal/statistics.py
"""Statistical utilities for Monte Carlo tally post-processing."""

import math


def compute_stats(batch_values: list[list[float]]) -> list[tuple[float, float, float]]:
    """Compute (mean, std_dev_of_mean, relative_error) for each bin across batches.

    Uses Welford's online algorithm for numerical stability.

    Args:
        batch_values: List of batches, each batch is a list of bin values.
    Returns:
        List of (mean, stddev, rel_error) tuples, one per bin.
    """
    if not batch_values:
        return []
    n_bins = len(batch_values[0])
    n_batches = len(batch_values)
    results = []

    for i in range(n_bins):
        mean = 0.0
        m2 = 0.0
        for b in range(n_batches):
            x = batch_values[b][i]
            delta = x - mean
            mean += delta / (b + 1)
            delta2 = x - mean
            m2 += delta * delta2

        variance = m2 / (n_batches - 1) if n_batches > 1 else 0.0
        std_dev = math.sqrt(variance / n_batches)  # std dev of mean
        rel_error = std_dev / abs(mean) if mean != 0.0 else 0.0
        results.append((mean, std_dev, rel_error))

    return results


def keff_statistics(batch_keff: list[float]) -> tuple[float, float, tuple[float, float]]:
    """Compute mean, std-dev-of-mean, and 95% CI for batch k-eff values.

    Returns: (mean, std_dev, (ci95_lower, ci95_upper))
    """
    n = len(batch_keff)
    if n == 0:
        return (0.0, 0.0, (0.0, 0.0))

    mean = 0.0
    m2 = 0.0
    for i in range(n):
        delta = batch_keff[i] - mean
        mean += delta / (i + 1)
        delta2 = batch_keff[i] - mean
        m2 += delta * delta2

    variance = m2 / (n - 1) if n > 1 else 0.0
    std_dev = math.sqrt(variance / n)
    ci95 = (mean - 1.96 * std_dev, mean + 1.96 * std_dev)
    return (mean, std_dev, ci95)


def figure_of_merit(rel_error: float, wall_time: float) -> float:
    """FOM = 1 / (R^2 * T). Higher is better."""
    if rel_error <= 0.0 or wall_time <= 0.0:
        return 0.0
    return 1.0 / (rel_error * rel_error * wall_time)


def batch_means(batch_values: list[list[float]], size: int) -> list[float]:
    """Compute mean across batches for each bin."""
    if not batch_values or size <= 0:
        return [0.0] * size
    sums = [0.0] * size
    for batch in batch_values:
        for i in range(min(size, len(batch))):
            sums[i] += batch[i]
    n = len(batch_values)
    return [s / n for s in sums]
